src/analysis/normed_space/dual.lean

```diff
@@ -115,3 +124,99 @@ def inclusion_in_double_dual_li : E →ₗᵢ[𝕜] (dual 𝕜 (dual 𝕜 E)) :=
 end bidual_isometry
 
 end normed_space
+
+section polar_sets
+
+open metric set normed_space
+
+/-- Given a subset `s` in a normed space `E` (over a field `𝕜`), the polar
+`polar 𝕜 s` is the subset of `dual 𝕜 E` consisting of those functionals which
+evaluate to something of norm at most one at all points `z ∈ s`. -/
+def polar (𝕜 : Type*) [nondiscrete_normed_field 𝕜]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] (s : set E) : set (dual 𝕜 E) :=
+{x' : dual 𝕜 E | ∀ z ∈ s, ∥ x' z ∥ ≤ 1 }
+
+namespace polar
+
+open metric set normed_space
+open_locale topological_space
+
+variables (𝕜 : Type*) [nondiscrete_normed_field 𝕜]
+variables {E : Type*} [normed_group E] [normed_space 𝕜 E]
+
+@[simp] lemma zero_mem (𝕜 : Type*) [nondiscrete_normed_field 𝕜]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] (s : set E) :
+  (0 : dual 𝕜 E) ∈ polar 𝕜 s :=
+λ _ _, by simp only [zero_le_one, continuous_linear_map.zero_apply, norm_zero]
+
+lemma eq_Inter (𝕜 : Type*) [nondiscrete_normed_field 𝕜]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] (s : set E) :
+  polar 𝕜 s = ⋂ z ∈ s, {x' : dual 𝕜 E | ∥ x' z ∥ ≤ 1 } :=
+by { ext, simp only [polar, mem_bInter_iff, mem_set_of_eq], }
+
+lemma of_empty (𝕜 : Type*) [nondiscrete_normed_field 𝕜]
+  {E : Type*} [normed_group E] [normed_space 𝕜 E] : polar 𝕜 (∅ : set E) = univ :=
+by { simp only [polar, forall_false_left, mem_empty_eq, forall_const, set_of_true], }
+
+variables {𝕜}
+
+/-- If `x'` is a dual element such that the norms `∥x' z∥` are bounded for `z ∈ s`, then a
+small scalar multiple of `x'` is in `polar 𝕜 s`. -/
+lemma smul_mem {s : set E} {x' : dual 𝕜 E} {c : 𝕜}
+  (hc : ∀ z, z ∈ s → ∥ x' z ∥ ≤ ∥c∥) : (c⁻¹ • x') ∈ polar 𝕜 s :=
+begin
+  by_cases c_zero : c = 0, { simp [c_zero] },
+  have eq : ∀ z, ∥ c⁻¹ • (x' z) ∥ = ∥ c⁻¹ ∥ * ∥ x' z ∥ := λ z, norm_smul c⁻¹ _,
+  have le : ∀ z, z ∈ s → ∥ c⁻¹ • (x' z) ∥ ≤ ∥ c⁻¹ ∥ * ∥ c ∥,
+  { intros z hzs,
+    rw eq z,
+    apply mul_le_mul (le_of_eq rfl) (hc z hzs) (norm_nonneg _) (norm_nonneg _), },
+  have cancel : ∥ c⁻¹ ∥ * ∥ c ∥ = 1,
+  by simp only [c_zero, norm_eq_zero, ne.def, not_false_iff,
+                inv_mul_cancel, normed_field.norm_inv],
+  rwa cancel at le,
+end
+
+/-- The `polar` of closed unit ball in a normed space `E` is the closed unit ball of the dual. -/
+lemma of_closed_unit_ball
+  {𝕜 : Type*} [is_R_or_C 𝕜] {E : Type*} [normed_group E] [normed_space 𝕜 E] :
+  polar 𝕜 (closed_ball (0 : E) 1) = closed_ball (0 : dual 𝕜 E) 1 :=
+begin
+  ext x',
+  simp only [mem_closed_ball, mem_set_of_eq, dist_zero_right],
+  split,
+  { intros h,
+    apply continuous_linear_map.op_norm_le_of_ball zero_lt_one zero_le_one,
+    intros z hz,
+    have key := linear_map.bound_of_ball_bound zero_lt_one 1 x'.to_linear_map h z,
```

Comment:
While trying to understand why `is_R_or_C` was needed on this lemma, I looked at the proof of `linear_map.bound_of_ball_bound`.  I wonder if actually you could give a proof of that lemma without using `is_R_or_C`?  My idea would be to use "rational approximation":  take rational multiples of `x` lying inside the ball.  The key point would seem to be
```lean
lemma norm_coe_rat (q : ℚ) : ∥(q : 𝕜)∥ = ↑|q| := sorry
```
I think this is true in an arbitrary normed field?

Reply:
I'm a bit puzzled. For example for `𝕜 := padic 3 _` (unfortunately I couldn't even get this to typecheck in Lean to test, because I don't know how to create a `fact` out of `nat.prime_three`), wouldn't the `∥(q : 𝕜)∥` on the left hand side be the 3-adic norm of `q` and `↑|q|` on the right hand side the usual absolute value of `q`? So the equality in the lemma would not hold? (Say `q=3` as a concrete counterexample.)

(Sorry, I am very much at loss with general `nondiscrete_normed_field`s... And I didn't manage to even state the above in Lean yet.)

Reply:
Sorry, of course you're right, ` ∥(q : 𝕜)∥ = ↑|q|` is not true in general!  Never mind.

... btw, regarding how to make a `fact` out of `nat.prime_three`, it's something like
```lean
lemma nat.fact_prime_three : fact (nat.prime 3) := ⟨nat.prime_three⟩

local attribute [instance] nat.fact_prime_three
```